lisa/_assets/kmodules/lisa/fs.c

```diff
@@ -0,0 +1,755 @@
+				ret = feature_param_remove_config_common(entry);
+			else
+				ret = feature_param_merge_common(entry);
+			if (ret) {
+				pr_err("Could not rollback config values\n");
+				return ret;
+			}
+		}
+	}
+
+	return ret;
+}
+
+static bool is_feature_set(char *name)
+{
+	struct feature_param_entry_value *val;
+
+	/* Check whether the feature is in the global set_features list. */
+	list_for_each_entry(val, &lisa_features_param.global_value, node)
+		if (lisa_features_param.ops->is_equal(&name, val))
+			return true;
+	return false;
+}
+
+int lisa_activate_config(bool value, struct lisa_cfg *cfg)
+{
+	struct feature *feature;
+	int ret;
+
+	if (cfg->activated == value)
+		return 0;
+
+	/* All the global values have now been updated. Time to enable them. */
+
+	ret = update_global_value(cfg, value);
+	if (ret)
+		return ret;
+
+	cfg->activated = value;
+
+	for_each_feature(feature) {
+		if (!is_feature_set(feature->name)) {
+			/*
+			 * Feature was enabled, and de-activating this config
+			 * disabled the feature.
+			 */
+			if (feature->__explicitly_enabled && !cfg->activated)
+				deinit_single_features(feature->name);
+			continue;
+		}
+
+		if (cfg->activated) {
+			/*
+			* Feature was enabled. By default, de-init before re-init the feature
+			* to catch potential modifications.
+			*/
+			if (feature->__explicitly_enabled)
+				deinit_single_features(feature->name);
+			init_single_feature(feature->name);
+			continue;
+		}
+	}
+
+	return 0;
+}
+
+static ssize_t lisa_activate_write(struct file *file, const char __user *buf,
+				   size_t count, loff_t *ppos)
+{
+	struct seq_file *s = file->private_data;
+	bool value;
+	int ret;
+
+	if (kstrtobool_from_user(buf, count, &value))
+		return -EINVAL;
+
+	mutex_lock(&interface_lock);
+	ret = lisa_activate_config(value, (struct lisa_cfg *)s->private);
+	mutex_unlock(&interface_lock);
+
+	return ret < 0 ? ret : count;
+}
+
+static int lisa_activate_open(struct inode *inode, struct file *file)
+{
+	struct lisa_cfg *cfg = inode->i_private;
+
+	return single_open(file, lisa_activate_show, cfg);
+}
+
+static struct file_operations lisa_activate_fops = {
+	.owner = THIS_MODULE,
+	.open = lisa_activate_open,
+	.read = seq_read,
+	.write = lisa_activate_write,
+	.release = single_release,
+};
+
+/*
+ * set_features handlers
+ * available_features handlers
+ */
+
+static void *lisa_param_feature_seq_start(struct seq_file *s, loff_t *pos)
+{
+	struct feature_param_entry *entry;
+	void *ret;
+
+	mutex_lock(&interface_lock);
+
+	entry = *(struct feature_param_entry **)s->private;
+	ret = seq_list_start(&entry->list_values, *pos);
+
+	return ret;
+}
+
+static int lisa_param_feature_seq_show(struct seq_file *s, void *v)
+{
+	struct feature_param_entry_value *val;
+	struct feature_param_entry *entry;
+	struct feature_param *param;
+
+	entry = *(struct feature_param_entry **)s->private;
+	param = entry->param;
+
+	val = hlist_entry(v, struct feature_param_entry_value, node);
+
+	if (param->ops->stringify) {
```

Comment:
What if the value is already a string ? Seems this is no being shown then ?

Reply:
But it's convenient to just have just one common case. I'll let you decide if a more optimal solution is required to handle strings.